Dockerfile

```diff
@@ -0,0 +1,137 @@
+    --mount=type=cache,target=/var/lib/apt/lists,target=/var/lib/apt/lists,sharing=locked,id=apt-lists \
+    apt-get install -y --no-install-recommends ${DEBIAN_BUILD_DEPS}
+
+
+# Alpine PHP with dependencies needed for building the tools
+FROM php-base-alpine AS builder-base-alpine
+ARG ALPINE_BUILD_DEPS="autoconf file g++ gcc libc-dev pkgconf re2c"
+RUN --mount=type=cache,target=/var/cache/apk,sharing=locked,id=apk \
+    apk add --no-cache ${ALPINE_BUILD_DEPS}
+
+
+# PHP with dependencies needed for building the tools
+FROM builder-base-${FLAVOUR} as builder-base
+
+
+# Stage containing the AST extension source
+FROM --platform=${BUILDPLATFORM} alpine AS ast-downloader
+WORKDIR /
+RUN --mount=type=cache,target=/var/cache/apk,sharing=locked,id=apk \
+    apk add --no-cache git
+RUN git clone https://github.com/nikic/php-ast.git
+
+
+# Stage that builds the AST extension from downloaded src
+FROM builder-base AS ast-builder
+WORKDIR /build
+COPY --from=ast-downloader /php-ast .
+RUN phpize  \
+    && ./configure  \
+    && make  \
+    && make install
+RUN docker-php-ext-enable ast
+
+
+# Stage that builds the pcov extension from PECL
+FROM builder-base AS pcov-builder
+RUN pecl install pcov && docker-php-ext-enable pcov
+
+
+# Stage that builds all other extensions
+FROM builder-base AS extensions-builder
+RUN docker-php-ext-install bcmath intl zip pcntl bz2
+
+
+# Stage containing the downloaded toolbox PHAR
+FROM --platform=${BUILDPLATFORM} alpine AS toolbox-downloader
+WORKDIR /
+ARG TOOLBOX_VERSION
+RUN wget https://github.com/jakzal/toolbox/releases/download/v$TOOLBOX_VERSION/toolbox.phar
+
+
+# PHP with all extensions needed, correctly configured
+FROM php-base as php-configured
+
+# Extensions .so
+COPY --link --from=ast-builder /usr/local/lib/php/extensions /usr/local/lib/php/extensions
+COPY --link --from=pcov-builder /usr/local/lib/php/extensions /usr/local/lib/php/extensions
+COPY --link --from=extensions-builder /usr/local/lib/php/extensions /usr/local/lib/php/extensions
+
+# Extensions .ini
+COPY --link --from=ast-builder /usr/local/etc/php/conf.d /usr/local/etc/php/conf.d
+COPY --link --from=pcov-builder /usr/local/etc/php/conf.d  /usr/local/etc/php/conf.d
+COPY --link --from=extensions-builder /usr/local/etc/php/conf.d /usr/local/etc/php/conf.d
+
+COPY --link <<EOF /usr/local/etc/php/conf.d/jakzal.ini
+date.timezone=Europe/London
+memory_limit=-1
+phar.readonly=0
+pcov.enabled=0
+EOF
+
+COPY --link --chmod=755 <<EOF /entrypoint.sh
+#!/usr/bin/env sh
+set -e
+
+(test "\$GITHUB_ACTIONS" = "true" || test "\$CI" = "true") && test -f composer.json && composer install --no-scripts --no-progress
+
+exec "\$@"
+EOF
+ARG TOOLBOX_TARGET_DIR="/tools"
+COPY --link --from=toolbox-downloader --chmod=+x /toolbox.phar ${TOOLBOX_TARGET_DIR}/toolbox
+COPY --link --from=composer:2 /usr/bin/composer ${TOOLBOX_TARGET_DIR}/composer
+
+# INSTALLATION_DATE is used to bust the docker layer cache
+ARG INSTALLATION_DATE
```

Comment:
I'd like to kill this, but the GH action we use doesn't support the needed flag yet https://github.com/docker/build-push-action/issues/645

Reply:
Ok. I thought it was smart 😆